counter/src/test/java/org/infinispan/counter/StrongCounterTest.java

```diff
@@ -96,6 +96,58 @@ public void testCompareAndSet(Method method) {
       }
    }
 
+   public void testCompareAndSetConcurrent(Method method) throws ExecutionException, InterruptedException, TimeoutException {
+      final int numThreadsPerNode = 8;
+      final int totalThreads = CLUSTER_SIZE * numThreadsPerNode;
+      final List<Future<LinkedList<Boolean>>> workers = new ArrayList<>(totalThreads);
+      final String counterName = method.getName();
+      final CyclicBarrier barrier = new CyclicBarrier(totalThreads);
+      final long maxIterations = 100;
+
+      for (int i = 0; i < totalThreads; ++i) {
+         final int cmIndex = i % CLUSTER_SIZE;
+         workers.add(fork(() -> {
+            long iteration = 0;
+            final long initialValue = 0;
+            long previousValue = initialValue;
+            LinkedList<Boolean> retValues = new LinkedList<>();
+            CounterManager manager = counterManager(cmIndex);
+            StrongTestCounter counter = createCounter(manager, counterName, initialValue);
+            while (iteration < maxIterations) {
+               assertEquals(previousValue, counter.getValue());
+               long update = previousValue + 1;
+               barrier.await();
+               //all threads calling compareAndSet at the same time, only one should succeed
+               boolean ret = counter.compareAndSet(previousValue, update);
+               if (ret) {
+                  previousValue = update;
+               } else {
+                  previousValue = counter.getValue();
+               }
+               retValues.add(ret);
```

Comment:
I was thinking that you don't need to store the result of every operation, something like this:

      final AtomicBoolean retValue = new AtomicBoolean(false);
      ...
               if (ret) {
                  assertTrue(retValue.compareAndSet(false, true));
               }
               barrier.await();
               retValue.set(false);

But this would only check that there's at most one successful operation, you still need something on the outside to check that there's at least one. So I was wrong, please go back to your original code.